vm/actor/src/builtin/puppet/mod.rs

```diff
@@ -0,0 +1,201 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use num_derive::FromPrimitive;
+use num_traits::FromPrimitive;
+
+use crate::check_empty_params;
+use address::Address;
+use encoding::{tuple::*, Cbor, CodecProtocol, Error};
+use ipld_blockstore::BlockStore;
+use num_bigint::bigint_ser;
+use runtime::{ActorCode, Runtime};
+use serde::{Deserialize, Serialize};
+use vm::{ActorError, ExitCode, MethodNum, Serialized, TokenAmount, METHOD_CONSTRUCTOR};
+
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    SendMethod = 2,
+    SendMarshalCBORFailure = 3,
+    ReturnMarshalCBORFailure = 4,
+    RuntimeTransactionMarshalCBORFailure = 5,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SendParams {
+    pub to: Address,
+    #[serde(with = "bigint_ser")]
+    pub value: TokenAmount,
+    pub method: u64,
+    pub params: Vec<u8>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SendReturn {
+    pub return_bytes: Serialized,
+    pub code: ExitCode,
+}
+
+#[derive(Default, Serialize, Deserialize)]
+pub struct FailToMarshalCBOR {}
+
+impl Cbor for FailToMarshalCBOR {
+    fn marshal_cbor(&self) -> Result<Vec<u8>, Error> {
+        Err(Error::Marshalling {
+            description: "Automatic fail to Marshall".to_string(),
+            protocol: CodecProtocol::Cbor,
+        })
+    }
+
+    fn unmarshal_cbor(_bz: &[u8]) -> Result<Self, Error> {
+        Err(Error::Unmarshalling {
+            description: "Automatic fail to Unmarshal".to_string(),
+            protocol: CodecProtocol::Cbor,
+        })
+    }
+}
+
+#[derive(Default, Serialize_tuple, Deserialize_tuple)]
+pub struct State {
+    opt_fail: Vec<Option<FailToMarshalCBOR>>,
+}
+
+impl Cbor for State {}
+
+pub struct Actor;
+
+impl Actor {
+    fn constructor<BS, RT>(rt: &mut RT) -> Result<(), ActorError>
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_accept_any()?;
+
+        rt.create(&State::default())?;
+        Ok(())
+    }
+
+    fn send<BS, RT>(rt: &mut RT, params: SendParams) -> Result<SendReturn, ActorError>
+    where
+        BS: BlockStore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_accept_any()?;
+
+        let res = rt.send(
+            params.to,
+            params.method,
+            Serialized::serialize(params.params).unwrap(),
+            params.value,
+        );
+
+        let return_bytes = res.map_err(|_| {
+            ActorError::new(
+                ExitCode::ErrIllegalState,
+                "failed to get return bytes".to_string(),
+            )
+        })?;
+
+        Ok(SendReturn {
+            return_bytes,
+            code: ExitCode::Ok,
+        })
+    }
+
+    fn send_failure<BS, RT>(rt: &mut RT, params: SendParams) -> Result<SendReturn, ActorError>
```

Comment:
I don't remember if i specifically look for this and its definitely a nitpick, but it'd be nice if we followed the name of these methods in specs-actor exactly